### DISCUSSION TITLE
Cost of Multiple Schedulers

### DISCUSSION THREAD

**Discussion by drusellers** [quartznet/quartznet#1802] [Q&A]
Broadly, what is the cost of using multiple schedulers?

```csharp
var scheduler = await schedulerFactory.GetScheduler("something");
```

Imagine an application taking a vertical slice / feature slice approach to an application. Would it make sense to have a scheduler per slice/feature? At what number of schedulers does it start to matter? If I have a max batch size of `X` does that limit apply to all jobs? Is the scheduler name mostly just another way to group jobs and triggers?

TY

**Comment by lahma** [comment]
> At what number of schedulers does it start to matter?

With default thread pool implementation which is Task-based, it shouldn't matter that much as system is constrained by the amount of concurrent Task handling by the CLR. Per scheduler there will be some maintenance threads, but they shouldn't matter that much until you have *many* schedulers. 

I haven't measured this so this is based mostly on my general thinking.

> If I have a max batch size of X does that limit apply to all jobs?

Batch acquisition happens on scheduler level, so you process batch size of *triggers* at a time.

> Is the scheduler name mostly just another way to group jobs and triggers?

It depends... It's a way for that for sure, but also allows different configurations. Like non-started scheduler using `ZeroSizeThreadPool` for doing only database maintenance. I don't have any numbers for this either. 

Usually there's only one scheduler per process, but as noted, there could be more. Having multiple schedulers in same process probably doesn't improve throughput as you can configure a "giant" scheduler with wanted concurrency levels allowing to process the amount of data that multiple schedulers would. For scaling, you would introduce more processing nodes to clustered scheduler.

I haven't thought about the feature slice that much. In my thinking jobs found be part of the feature and could share the same scheduler, using different job/trigger groups maybe derived from feature.

**Comment by drusellers** [comment]
> It depends... It's a way for that for sure, but also allows different configurations.

Using the `services.AddQuartz` model - what would that look like?

> Having multiple schedulers in same process probably doesn't improve throughput

Yeah, I didn't think it would. Feels more like an organizational tool to me at this point in my understanding.

> I haven't thought about the feature slice that much. In my thinking jobs found be part of the feature and could share the same scheduler, using different job/trigger groups maybe derived from feature.

Having a scheduler per feature slice that I can start/stop as a whole seems nice. Then inside of the feature I can have groups and triggers.

**Reply by drusellers** [reply]
> Batch acquisition happens on scheduler level, so you process batch size of triggers at a time

So if I set the MaxBatchSize to `10` and I have two schedulers, I'll be working through 20 triggers at a time?

**Reply by lahma** [reply]
> > It depends... It's a way for that for sure, but also allows different configurations.
> 
> Using the `services.AddQuartz` model - what would that look like?

Currently I think it's impossible with the API, one would need probably overload to either supply scheduler name or factory interface as parameter. Like `AddQuartz("userModuleScheduler", options => )` but it has the problem that then it would be quite impossible to get `ISchedulerFactory` from standard MS DI (I think.). I like the approach that Marten did for multiple `IDocumentStore` instances via inheritance. Then in Quartz it would be `AddQuartz<IUserModuleSchedulerFactory>(options => )` (`where T : ISchedulerFactory`). But no API surface (yet). It's not a common thing so far. 

One could probably utilize `SchdulerBuilder` for workaround too, it's what the API uses under the hood anyway, which then actually just populates `NameValueCollection`.

> > Batch acquisition happens on scheduler level, so you process batch size of triggers at a time

> So if I set the MaxBatchSize to 10 and I have two schedulers, I'll be working through 20 triggers at a time?

Max 10 per scheduler yes, depends on how much there are waiting in the database. Two schedulers can share DB schema as scheduler name is part of the final key for retrieval.

**Reply by drusellers** [reply]
> Currently I think it's impossible with the API

Totally understand, and is what I thought. TY
